dev/buildtool/spin_commands.py

```diff
@@ -436,6 +398,49 @@ def init_argparser(self, parser, defaults):
     BomSourceCodeManager.add_parser_args(parser, defaults)
 
 
+def bump_spin_patch(git, git_dir, gate_version):
+  '''Calculates the next spin version from the gate version and previous spin tags.
+
+  Spin is coupled to the Gate major and minor version.
+  Gate is a routing server, so features and breaking changes in Gate
+  must be reflected in spin since it is a client.
+
+  :param git: git support helper class.
+  :param git_dir: spin git directory.
+  :param gate_version: gate version to align spin version with in <maj>.<min>.<patch>-<buildnum> format.
+  :return: (SemanticVersion) Next semver spin version.
+  '''
+  gate_version_parts = gate_version.split('-')
+  if len(gate_version_parts) != 2:
+    raise_and_log_error(
+        ValueError('Malformed gate version {}'.format(gate_version)))
+
+  # SemanticVersion.make() expects a tag, so formulate the input gate version as a tag.
+  gate_semver = SemanticVersion.make('version-{}'.format(gate_version_parts[0]))
+  tag_pattern = r'version-{maj}.{min}.(\d+)'.format(maj=gate_semver.major,
+                                                    min=gate_semver.minor)
+  tag_matcher = re.compile(tag_pattern)
+  tags = git.fetch_tags(git_dir)
+
+  logging.info('searching git tags {} for patterns matching {}'.format(tags, tag_pattern))
+  matching_semvers = [SemanticVersion.make(t) for t in tags if tag_matcher.match(t)]
+  logging.info('found matching semvers: {}'.format(matching_semvers))
+  if matching_semvers:
+    max_semver = max(matching_semvers)
+    next_semver = max_semver.next(SemanticVersion.PATCH_INDEX)
+    patch = next_semver.patch
```

Comment:
general question -- how do you know if you even need a new version and this is different from some previous one? Browsing the broader file this wasnt clear to me.

Reply:
`bump_spin_patch` is just the helper. We call this in two places:
1. When we record the 'internal version' the spin binary will report when asked. We always set this internal version as the Gate version + next patch if no tags are present. This is OK because we won't publish every spin binary built here, see (2).
2. When we publish the spin binary. We check if the locally checked-out HEAD commit is different from the commit at the most previous tag and roll up the patch if they aren't.

I agree it's a bit confusing and should be clarified in comments.

Reply:
I think the __no_changes at #308 is what I was looking for.